File: leetCode/Python2/22-generateParenthese.py
class Solution(object):
    def __init__(self):
        self.res = []
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        self.res = []
        self.parenthese_helper(n, n, "")
        return self.res

    def parenthese_helper(self, left, right, parenthese):
        if left > right:
            return
        if left == 0 and right == 0:
            self.res.append(parenthese)
            return
        if left > 0:
            self.parenthese_helper(left - 1, right, parenthese + '(')
        if right > 0:
            self.parenthese_helper(left, right - 1, parenthese + ')')

# good reference: https://zh.wikipedia.org/wiki/%E5%8D%A1%E5%A1%94%E5%85%B0%E6%95%B0

if __name__ == "__main__":
    s = Solution()
    print s.generateParenthesis(2)
    print s.generateParenthesis(3)

File: leetCode/C++/22-generateParenthese.cc
/*
Given n pairs of parentheses, write a function to generate
all combinations of well-formed parentheses.

For example, given n = 3, a solution set is:

"((()))", "(()())", "(())()", "()(())", "()()()"

*/

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        if (n > 0)
            parenthesis_generator(n, n, "", res);
        return res;
    }
private:
    void parenthesis_generator(int left, int right, string parent, vector<string>& res) {
        if (left > right)
            return;
        if (left == 0 && right == 0) {
            res.push_back(parent);
            return;
        }
        if (left > 0)
            parenthesis_generator(left - 1, right, parent + "(", res);
        if (right > 0)
            parenthesis_generator(left, right - 1, parent + ")", res);
    }
};

int main(void) {
    Solution s;
    vector<string> res = s.generateParenthesis(3);
    for (auto const &parent: res)
        cout << parent << " ";
    cout << endl;
    return 0;
};